### PR TITLE
Alert for reactive slacks over a threshold value

### PR DESCRIPTION
**Pull request by FranckLecuyer**

### CODE REVIEW

**Review by TheMaskedTurtle — CHANGES_REQUESTED** [review]
As discussed I'm not convinced by the notification architecture here. We're adding really specific headers, and sending them through a result message. The study-server will just pass them, then we have to define these new headers everywhere. It could be more generic and easier to send a generic event directly to the study-notification-server.

**Comment by FranckLecuyer** [comment]
> As discussed I'm not convinced by the notification architecture here. We're adding really specific headers, and sending them through a result message. The study-server will just pass them, then we have to define these new headers everywhere. It could be more generic and easier to send a generic event directly to the study-notification-server.

For me, this should be done in another US, to handle also the computation statuses of all computation models.
The notifications must be displayed only for the study concerned, and it's the study-server that actually knows this information ...
Sending event directly to the study-notification-server implies that all the computation servers must have this study uuid ...

File: src/main/java/org/gridsuite/voltageinit/server/service/VoltageInitWorkerService.java [hunk 1]
```diff
@@ -128,30 +131,31 @@ public static void addRestrictedVoltageLevelReport(Map<String, Double> voltageLe
...
+                .withKey("reactiveSlacksOverThreshold")
+                .withDefaultMessage("Reactive slack exceeds ${threshold} MVar for at least one bus")
+                .withValue("threshold", reactiveSlacksThreshold.get())
+                .withSeverity(TypedValue.WARN_SEVERITY)
+                .build());
+            result = Pair.of(Boolean.TRUE, reactiveSlacksThreshold.get());
+        }
+        return result;
+    }
+
```
[the remark this answers is not in the document]
Comment (FranckLecuyer) [reply src/main/java/org/gridsuite/voltageinit/server/service/VoltageInitWorkerService.java]: Done

File: src/main/java/org/gridsuite/voltageinit/server/service/VoltageInitService.java [hunk 2]
```diff
@@ -39,6 +39,8 @@ public class VoltageInitService {
 
     private final VoltageInitResultRepository resultRepository;
 
+    public static final double DEFAULT_REACTIVE_SLACKS_THRESHOLD = 500.;
```
[the remark this answers is not in the document]
Comment (FranckLecuyer) [reply src/main/java/org/gridsuite/voltageinit/server/service/VoltageInitService.java]: Done

File: src/main/java/org/gridsuite/voltageinit/server/service/VoltageInitWorkerService.java [hunk 3]
```diff
@@ -204,8 +208,22 @@ public Consumer<Message<String>> consumeRun() {
...
+                AtomicReference<Reporter> rootReporter = new AtomicReference<>(Reporter.NO_OP);
+                Reporter reporter = Reporter.NO_OP;
+                if (context.getReportUuid() != null) {
+                    String rootReporterId = context.getReporterId() == null ? VOLTAGE_INIT_TYPE_REPORT : context.getReporterId() + "@" + context.getReportType();
+                    rootReporter.set(new ReporterModel(rootReporterId, rootReporterId));
+                    reporter = rootReporter.get().createSubReporter(context.getReportType(), VOLTAGE_INIT_TYPE_REPORT, VOLTAGE_INIT_TYPE_REPORT, context.getReportUuid().toString());
+                    // Delete any previous VoltageInit computation logs
+                    voltageInitObserver.observe("report.delete", () ->
+                        reportService.deleteReport(context.getReportUuid(), context.getReportType()));
+                }
```
[the remark this answers is not in the document]
Comment (FranckLecuyer) [reply src/main/java/org/gridsuite/voltageinit/server/service/VoltageInitWorkerService.java]: Done

File: src/main/java/org/gridsuite/voltageinit/server/service/VoltageInitWorkerService.java [hunk 4]
```diff
@@ -214,19 +232,34 @@ public Consumer<Message<String>> consumeRun() {
                 if (openReacResult != null) {  // result available
                     UUID modificationsGroupUuid = createModificationGroup(openReacResult, network);
                     Map<String, Bus> networkBuses = network.getBusView().getBusStream().collect(Collectors.toMap(Bus::getId, Function.identity()));
+                    // check if at least one reactive slack over the threshold value
+                    Pair<Boolean, Double> resultCheckReactiveSlacks = checkReactiveSlacksOverThreshold(openReacResult, context.getParametersUuid(), reporter);
+                    boolean isReactiveSlacksOverThreshold = resultCheckReactiveSlacks != null ? resultCheckReactiveSlacks.getLeft() : Boolean.FALSE;
+                    Double reactiveSlacksOverThreshold = resultCheckReactiveSlacks != null ? resultCheckReactiveSlacks.getRight() : null;
```
[the remark this answers is not in the document]
Comment (FranckLecuyer) [reply src/main/java/org/gridsuite/voltageinit/server/service/VoltageInitWorkerService.java]: Done

File: src/main/java/org/gridsuite/voltageinit/server/service/NotificationService.java [hunk 5]
```diff
@@ -41,6 +41,9 @@ public class NotificationService {
     public static final String HEADER_RECEIVER = "receiver";
     public static final String HEADER_MESSAGE = "message";
     public static final String HEADER_USER_ID = "userId";
+    public static final String HEADER_REACTIVE_SLACKS_OVER_THRESHOLD = "REACTIVE_SLACKS_OVER_THRESHOLD";
+    public static final String HEADER_REACTIVE_SLACKS_THRESHOLD_VALUE = "reactiveSlacksThreshold";
+    public static final String REACTIVE_SLACKS_OVER_THRESHOLD = "REACTIVE_SLACKS_OVER_THRESHOLD";
```
Comment (TheMaskedTurtle) [inline src/main/java/org/gridsuite/voltageinit/server/service/NotificationService.java:46]: This is unused

**Review by TheMaskedTurtle — APPROVED** [review]

File: src/main/java/org/gridsuite/voltageinit/server/service/NotificationService.java (hunk 5 above)
Comment (FranckLecuyer) [reply src/main/java/org/gridsuite/voltageinit/server/service/NotificationService.java]: Removed

File: src/main/java/org/gridsuite/voltageinit/server/service/parameters/VoltageInitParametersService.java [hunk 6]
```diff
@@ -73,10 +75,18 @@ public Optional<UUID> duplicateParameters(UUID sourceParametersId) {
...
+    @Transactional
     public VoltageInitParametersInfos getParameters(UUID parametersUuid) {
         return voltageInitParametersRepository.findById(parametersUuid).map(VoltageInitParametersEntity::toVoltageInitParametersInfos).orElse(null);
     }
 
+    @Transactional
+    public double getReactiveSlacksThreshold(UUID parametersUuid) {
+        return Optional.ofNullable(parametersUuid != null ? voltageInitParametersRepository.findById(parametersUuid).map(VoltageInitParametersEntity::toVoltageInitParametersInfos).orElse(null) : null)
+            .map(VoltageInitParametersInfos::getReactiveSlacksThreshold)
+            .orElse(DEFAULT_REACTIVE_SLACKS_THRESHOLD);
```
Comment (TheMaskedTurtle) [inline src/main/java/org/gridsuite/voltageinit/server/service/parameters/VoltageInitParametersService.java:87]:
```suggestion
        return Optional.ofNullable(parametersUuid)
            .flatMap(voltageInitParametersRepository::findById)
            .map(VoltageInitParametersEntity::toVoltageInitParametersInfos)
            .map(VoltageInitParametersInfos::getReactiveSlacksThreshold)
            .orElse(DEFAULT_REACTIVE_SLACKS_THRESHOLD);
```

**Review by TheMaskedTurtle — APPROVED** [review]

File: src/main/java/org/gridsuite/voltageinit/server/service/parameters/VoltageInitParametersService.java (hunk 6 above)
Comment (FranckLecuyer) [reply src/main/java/org/gridsuite/voltageinit/server/service/parameters/VoltageInitParametersService.java]: Done

**Pull request merged**